File: ios/CatchMeUp/Engine/Exam.swift
import Foundation

/// Practice questions from a brain's recaps. Offline — same idea as
/// `./catchup exam`: terms, key moments, and study prompts, graded by overlap.
struct ExamQuestion: Identifiable, Hashable {
    var id = UUID()
    var prompt: String
    var answer: String
    var kind: Kind
    var source: String
    var timestamp: String
    var concept: String
    var keys: [String]

    enum Kind: String, Hashable {
        case term, moment, notes, study
    }
}

struct ExamGrade: Hashable {
    var score: Double
    var verdict: Verdict
    var because: String

    enum Verdict: String {
        case pass, partial, miss, blank

        var title: String {
            switch self {
            case .pass: return "Got it"
            case .partial: return "Partly"
            case .miss: return "Not yet"
            case .blank: return "Blank"
            }
        }
    }
}

enum ExamBuilder {
    private static let stop: Set<String> = [
        "the", "and", "for", "that", "this", "with", "from", "are", "was",
        "what", "when", "your", "you", "not", "but", "use", "used", "using",
        "can", "will", "how", "why", "into", "about", "have", "has",
    ]

    static func build(from recordings: [Recording], count: Int = 8) -> [ExamQuestion] {
        var pool: [ExamQuestion] = []
        var seen = Set<String>()

        func add(_ q: ExamQuestion?) {
            guard let q, !q.prompt.isEmpty, !q.answer.isEmpty else { return }
            let key = q.concept.lowercased()
            guard seen.insert(key).inserted else { return }
            pool.append(q)
        }

        for rec in recordings {
            guard let recap = rec.recap else { continue }
            let source = rec.displayTitle
            for term in recap.terms ?? [] {
                add(termQuestion(term, source: source))
            }
            for bm in recap.bookmarks ?? [] {
                add(momentQuestion(bm, source: source))
            }
            for note in recap.detailedNotes ?? [] {
                add(notesQuestion(note, source: source))
            }
            for item in recap.study ?? [] {
                add(studyQuestion(item, source: source))
            }
        }

        guard !pool.isEmpty else { return [] }
        var byKind: [ExamQuestion.Kind: [ExamQuestion]] = [
            .term: [], .moment: [], .notes: [], .study: [],
        ]
        for q in pool { byKind[q.kind, default: []].append(q) }
        for kind in byKind.keys { byKind[kind]?.shuffle() }

        let n = max(1, count)
        let quota: [(ExamQuestion.Kind, Int)] = [
            (.term, max(1, n * 2 / 5)),
            (.moment, max(1, n * 2 / 5)),
            (.notes, max(0, n / 5)),
            (.study, max(0, n / 5)),
        ]
        var picked: [ExamQuestion] = []
        var seenPrompt = Set<String>()
        func take(_ q: ExamQuestion) -> Bool {
            guard seenPrompt.insert(q.prompt.lowercased()).inserted else { return false }
            picked.append(q)
            return true
        }
        for (kind, want) in quota {
            var got = picked.filter { $0.kind == kind }.count
            while got < want, let next = byKind[kind]?.popLast(), picked.count < n {
                if take(next) { got += 1 }
            }
        }
        var leftover = byKind.values.flatMap { $0 }
        leftover.shuffle()
        for q in leftover where picked.count < n { _ = take(q) }
        picked.shuffle()
        return Array(picked.prefix(n))
    }

    static func grade(_ question: ExamQuestion, typed: String) -> ExamGrade {
        let typedTrim = typed.trimmingCharacters(in: .whitespacesAndNewlines)
        if typedTrim.isEmpty {
            return ExamGrade(score: 0, verdict: .blank, because: "empty answer")
        }
        let gold = question.answer.lowercased()
        let got = typedTrim.lowercased()
        if gold.contains(got) || got.contains(gold) {
            return ExamGrade(score: 1, verdict: .pass, because: "matches the lecture wording")
        }
        let keys = question.keys.isEmpty ? tokens(question.answer).filter { !stop.contains($0) } : question.keys
        let typedTokens = Set(tokens(typedTrim))
        let hits = keys.filter { key in
            let k = key.lowercased()
            return got.contains(k) || typedTokens.contains(k)
        }
        let denom = max(1, min(keys.count, 5))
        let score = Double(hits.count) / Double(denom)
        var because = hits.isEmpty ? "" : "you hit: " + hits.prefix(4).joined(separator: ", ")
        let missed = keys.filter { !hits.contains($0) }.prefix(3)
        if !missed.isEmpty {
            because += (because.isEmpty ? "" : " · ") + "still needed: " + missed.joined(separator: ", ")
        }
        let verdict: ExamGrade.Verdict
        if score >= 0.45 && hits.count >= 2 { verdict = .pass }
        else if score >= 0.25 || hits.contains(where: { $0.count > 4 }) { verdict = .partial }
        else { verdict = .miss }
        return ExamGrade(score: (score * 100).rounded() / 100, verdict: verdict, because: because)
    }

    static func tokens(_ text: String) -> [String] {
        text.lowercased()
            .split { !$0.isLetter && !$0.isNumber }
            .map(String.init)
            .filter { $0.count > 2 }
    }

    private static func keys(from gold: String, extra: String = "") -> [String] {
        var out: [String] = []
        for w in tokens(gold + " " + extra) where !stop.contains(w) {
            if !out.contains(w) { out.append(w) }
            if out.count >= 6 { break }
        }
        return out
    }

    private static func firstSentences(_ text: String, limit: Int = 200) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }
        let parts = trimmed.split { ".!?".contains($0) }.map {
            $0.trimmingCharacters(in: .whitespacesAndNewlines)
        }.filter { !$0.isEmpty }
        var out = parts.prefix(2).joined(separator: ". ")
        if out.count > limit {
            out = String(out.prefix(limit - 1)).trimmingCharacters(in: .whitespaces) + "…"
        }
        return out
    }

    private static func termQuestion(_ term: Term, source: String) -> ExamQuestion? {
        let gold = firstSentences(term.definition)
        guard gold.count >= 20 else { return nil }
        let shown = term.term.trimmingCharacters(in: .whitespaces)
        guard shown.count >= 2 else { return nil }
        let keyList = keys(from: gold, extra: shown)
        guard keyList.count >= 2 else { return nil }
        return ExamQuestion(
            prompt: "What is \(indefinite(shown))? One or two sentences, in your own words.",
            answer: gold,
            kind: .term,
            source: source,
            timestamp: "",
            concept: shown.lowercased(),
            keys: keyList
        )
    }

    private static func momentQuestion(_ bm: Bookmark, source: String) -> ExamQuestion? {
        let gold = firstSentences(bm.insight)
        guard gold.count >= 20, !bm.heading.isEmpty else { return nil }
        return ExamQuestion(
            prompt: "Why does “\(bm.heading)” matter? One or two sentences.",
            answer: gold,
            kind: .moment,
            source: source,
            timestamp: bm.timestamp,
            concept: bm.heading.lowercased(),
            keys: keys(from: gold, extra: bm.heading)
        )
    }

    private static func notesQuestion(_ note: DetailNote, source: String) -> ExamQuestion? {
        let gold = firstSentences(note.content)
        guard gold.count >= 24, !note.heading.isEmpty else { return nil }
        return ExamQuestion(
            prompt: "Explain \(note.heading) the way the lecture did.",
            answer: gold,
            kind: .notes,
            source: source,
            timestamp: "",
            concept: note.heading.lowercased(),
            keys: keys(from: gold, extra: note.heading)
        )
    }

    private static func studyQuestion(_ item: String, source: String) -> ExamQuestion? {
        let text = item.trimmingCharacters(in: .whitespacesAndNewlines)
        guard text.count >= 16 else { return nil }
        let prompt = text.hasSuffix("?") ? text : text
        return ExamQuestion(
            prompt: prompt,
            answer: firstSentences(item, limit: 240),
            kind: .study,
            source: source,
            timestamp: "",
            concept: String(text.lowercased().prefix(80)),
            keys: keys(from: item)
        )
    }

    private static func indefinite(_ term: String) -> String {
        let t = term.lowercased()
        guard t.first?.isLetter == true, !t.contains("["), t.split(separator: " ").count <= 4 else {
            return term
        }
        let vowels = "aeiou"
        let article = vowels.contains(t.first!) ? "an" : "a"
        return "\(article) \(t)"
    }
}

File: exam.py
#!/usr/bin/env python3
"""Practice exams from a CatchMeUp brain (terms, study list, cortex gaps)."""
from __future__ import annotations

import argparse
import random
import sys
from pathlib import Path

PROJECT_DIR = Path(__file__).resolve().parent
if str(PROJECT_DIR) not in sys.path:
    sys.path.insert(0, str(PROJECT_DIR))

import brains
import library
from brains import _tokens


def _records(brain: str | None):
    rows = library.list_records(mode="lecture", brain=brain) or library.list_records(brain=brain)
    return rows


def build_exam(records: list[dict], count: int = 8, rng: random.Random | None = None) -> list[dict]:
    """Build short-answer questions from recaps. No network."""
    rng = rng or random.Random()
    pool: list[dict] = []
    seen = set()

    def add(prompt: str, answer: str, source: str, kind: str, timestamp: str = ""):
        key = prompt.strip().lower()
        if not prompt.strip() or not answer.strip() or key in seen:
            return
        seen.add(key)
        pool.append({
            "kind": kind,
            "prompt": prompt.strip(),
            "answer": answer.strip(),
            "source": source,
            "timestamp": timestamp,
        })

    for rec in records:
        analysis = rec.get("analysis") or {}
        title = rec.get("title") or rec.get("source") or "recap"
        for term, definition in library._terms_from(rec):
            if definition.startswith("(from the study"):
                add(term, f"(from {title}) {term}", title, "study")
            else:
                add(f"Define: {term}", definition, title, "term")
        for bm in analysis.get("bookmarks") or []:
            if not isinstance(bm, dict):
                continue
            heading = (bm.get("heading") or "").strip()
            insight = (bm.get("insight") or "").strip()
            if heading and insight:
                add(
                    f"Why does “{heading}” matter?",
                    insight,
                    title,
                    "moment",
                    str(bm.get("timestamp") or ""),
                )
        for section in analysis.get("detailed_notes") or []:
            if isinstance(section, dict) and section.get("heading") and section.get("content"):
                add(
                    f"Explain: {section['heading']}",
                    str(section["content"])[:400],
                    title,
                    "notes",
                )
    rng.shuffle(pool)
    return pool[: max(1, count)] if pool else []


def grade_answer(question: dict, typed: str) -> dict:
    """Cheap token overlap — used when you don't want to spend an LLM call."""
    expected = set(_tokens(question.get("answer") or ""))
    got = set(_tokens(typed or ""))
    if not typed.strip():
        return {"score": 0.0, "verdict": "blank", "overlap": 0}
    if not expected:
        return {"score": 0.5, "verdict": "ungraded", "overlap": 0}
    overlap = len(expected & got)
    score = overlap / max(1, len(expected))
    if score >= 0.45 or (typed.strip() and typed.strip().lower() in (question.get("answer") or "").lower()):
        verdict = "pass"
    elif score >= 0.2:
        verdict = "partial"
    else:
        verdict = "miss"
    return {"score": round(score, 2), "verdict": verdict, "overlap": overlap}


def format_exam(questions: list[dict], answers: bool = False) -> str:
    if not questions:
        return "No exam material yet. Recap a lecture into this brain first."
    lines = [f"CatchMeUp exam — {len(questions)} question(s)", ""]
    for i, q in enumerate(questions, 1):
        where = q.get("source") or ""
        ts = f"  [{q['timestamp']}]" if q.get("timestamp") else ""
        lines.append(f"{i}. {q['prompt']}")
        lines.append(f"    ({q.get('kind')}) {where}{ts}")
        if answers:
            lines.append(f"    → {q['answer']}")
        lines.append("")
    if not answers:
        lines.append("Reveal answers:  ./catchup exam NAME --print --answers")
        lines.append("Take it live:    ./catchup exam NAME")
        lines.append("Hear a moment:   ./catchup clip NAME <words from the question>")
    return "\n".join(lines).rstrip() + "\n"


def cmd_exam(brain: str | None, count: int, print_only: bool, show_answers: bool):
    if brain:
        brains.load_brain(brain)
    rows = _records(brain)
    if not rows:
        print("Nothing to examine. Recap a lecture first:")
        print("  ./catchup lecture FILE")
        print("  ./catchup into cs61a FILE")
        sys.exit(1)
    questions = build_exam(rows, count=count)
    if not questions:
        print("Recaps have no terms / study items / notes to examine yet.")
        sys.exit(1)
    if print_only:
        print(format_exam(questions, answers=show_answers))
        return
    print(f"Exam — {len(questions)} questions from your recaps. q = quit.\n")
    passed = 0
    attempted = 0
    import viz

    for i, q in enumerate(questions, 1):
        print(f"[{i}/{len(questions)}]  {q['prompt']}")
        print(f"         {q.get('source')}")
        try:
            typed = input("         your answer: ").strip()
        except EOFError:
            print()
            break
        if typed.lower() in {"q", "quit"}:
            break
        attempted += 1
        result = grade_answer(q, typed)
        mark = viz.verdict_mark(result["verdict"])
        print(f"         {mark}  {q['answer'][:220]}")
        if result["verdict"] == "pass":
            passed += 1
        print()
    print(viz.exam_card(passed, attempted or 0, len(questions)))
    if brain:
        print(f"Go deeper:  ./catchup think {brain} <the one you missed>")
        print(f"Hear it:    ./catchup clip {brain} <term>")


def main(argv=None):
    parser = argparse.ArgumentParser(prog="catchup-exam")
    parser.add_argument("--brain")
    parser.add_argument("-n", "--count", type=int, default=8)
    parser.add_argument("--print", dest="print_only", action="store_true")
    parser.add_argument("--answers", action="store_true")
    args = parser.parse_args(argv)
    cmd_exam(args.brain, args.count, args.print_only, args.answers)


if __name__ == "__main__":
    main()
